### ISSUE TITLE
Electron + Webpack + React: SQLite package has not been found installed

### ISSUE THREAD

**Issue by rohmanhm** [typeorm/typeorm#2456]
**Issue type:**

[x] question
[ ] bug report
[ ] feature request
[ ] documentation issue

**Database system/driver:**

[ ] `cordova`
[ ] `mongodb`
[ ] `mssql`
[ ] `mysql` / `mariadb`
[ ] `oracle`
[ ] `postgres`
[x] `sqlite`
[ ] `sqljs`
[ ] `react-native`

**TypeORM version:**

[x] `latest`
[ ] `@next`
[ ] `0.x.x` (or put your version here)

**Steps to reproduce or a small repository showing the problem:**

I using this boilerplate to make my electron app. https://github.com/iRath96/electron-react-typescript-boilerplate

I follow the same instruction as https://github.com/typeorm/electron-typescript-example did, but when I run `yarn dev` on my project it throws an error like below on console developer tools electron app.

```bash
Error:  Error: SQLite package has not been found installed. Try to install it: npm install sqlite3 --save
    at new DriverPackageNotInstalledError 
```

![image](https://user-images.githubusercontent.com/7524911/42227443-be4fa49c-7f0b-11e8-809b-16f5015b554b.png)

I have removed and reinstall node_modules but it didn't work.
I also double check if `sqlite3` present on node_modules or not. But it's available.

![image](https://user-images.githubusercontent.com/7524911/42227531-eb99d97c-7f0b-11e8-9902-6ff3aa449e92.png)

So I've made repository if you want to check it out, here is my repo https://github.com/rohmanhm/ts-electron-react-typeorm-sqlite

Any way to solve this issue? Any suggestion will be helpful.

**Comment by HyperSprite** [comment]
This my help, maybe it won't...
I am using electron-react-boilerplate (no typescript) and using TypeORM from main with ipc coms from render with SQLite (I was using SQL.js but TypeORM seems does not expect that in node, that worked in dev but I could never get it to build in prod no matter how I installed things). Also, not using decorators so I feel like I'm doing everything the hard way.
In order for it to work correctly for hot Dev and Prod build, I had to 

* Install SQLite and TypeORM in the app/package.json, even though TypeORM is not native. 
* Do a rebuild after the install of SQLite, even though it was supposed to do that automatically. 
* Move my entity folder and ormconfig into app/
* Import the ormconfig file directly to the createConnection(). 

I may be doing this all wrong but TypeORM seems to make assumptions about where things are supposed to be installed (near the package.json file, lol) and this was the only way I could get it to work.
Here is my working repo: https://github.com/HyperSprite/electron-react-boilerplate-with-material-ui

**Comment by rohmanhm** [comment]
Thanks for your solution @HyperSprite, maybe that's a little complicated.
I'll take a look your repo. 

I'm still waiting for @pleerock , maybe he has a better solution for this issue.

**Comment by pleerock** [comment]
> I also double check if sqlite3 present on node_modules or not. But it's available.

There is a `catch` block that catches error and gives you another error. Try to console.log original error in the SqliteDriver source code and see what exactly error you have.

> I follow the same instruction as https://github.com/typeorm/electron-typescript-example did,

make sure to do exact instructions, also use npm instead of yarn just to be sure that its not related to yarn somehow.

Usually problem can be with **native module** built by sqlite make sure its there (post-install script does it) and its built for electron.

**Comment by rohmanhm** [comment]
I removed and reinstall all the packages using `npm install` and run the command using `npm`, But the result still the same.

This is what error I've got after console.log from SqliteDriver source code.
```bash
Cannot find module 'C:\Users\Romsy\Documents\Code\Projects\savings\node_modules\sqlite3\lib\binding\electron-v1.8-win32-x64\node_sqlite3.node'
    at Module._resolveFilename 
```

I think it's related to https://github.com/electron/electron/issues/10220

**Comment by rohmanhm** [comment]
I finally fixed my issue by installing this package https://github.com/electron/electron-rebuild, and run
```bash
electron-rebuild -f -w sqlite3
```

Thanks, everyone 🙂

**Comment by rohmanhm** [comment]
Maybe this one needs to be added to the docs since a lot of windows users like me getting the same issue.

**Comment by pleerock** [comment]
there is actually a post-install script in package.json that runs electron-rebuild for you

**Comment by dbishoponline** [comment]
I have same issue.  I have tried using electron-rebuild, but no luck.

**Comment by HyperSprite** [comment]
I know this is old and closed but just want to add a note about my final solution with Electron. Not sure it will work in this case but thought I'd add this anyway. 

The real problem for me was the Electron React Boilerplate (ERB) was using a 2 package.json file system and even though I was able to get things to work initially, it was whack-a-mole with other issues. I ended up collapsing ERB into a single package.json repo (I filed a PR back to ERB last year, it might finally be in their .next branch now, not sure, my project is no longer tracking to that). The single package.json system makes the electron-builder work as expected without any extra config or messing around.

**Comment by jjhbw** [comment]
I ran into the same problem (Electron + Create React App / Webpack).
For me the only solution ended up being https://github.com/typeorm/typeorm/issues/4210#issuecomment-559401079

**Comment by ItsMeSousuke** [comment]
It's working on 0.2.13, on newer version it's not working and show error like in issue. Try to downgrade typeorm dependency to this version.

**Comment by darekf77** [comment]
For my the fix was to change install script from:
```
npm install --ignore-scripts
```
to
```
npm install
```

**Comment by oktapodia** [comment]
Just a follow up for recent versions.

add sqlite3 to the dependencies of the `release/app/package.json` file

pass manually the driver (typeorm is doing a relative path resolution)

```
import sqlite3 from 'sqlite3';
    const dataSource = new DataSource({
      type: "sqlite",
      database: "database.sqlite",
      driver: sqlite3, // The important line
      synchronize: true,
      logging: true,
      entities: [User],
      migrations: [],
      subscribers: [],
    })
```

No errors anymore

**Comment by vijaysharma0757** [comment]
I have also faced same issue. Resolved by several steps.
Steps
1. Add sqlite and typeorm in `release/app/package.json` file. Please install only version for sqlite@5.1.6
`npm install sqlite3@5.1.6 --save` 
`npm install typeorm --save`
2.  Remove `sqlite` and `typeorm` from `dependencies` in `package.json` file and add to `devDependencies`
`npm install sqlite3@5.1.6 --save-dev`
`npm install typeorm --save-dev`

Below error resolved by changing sqlite version `5.1.6`
<img width="1255" alt="Screenshot 2024-04-09 at 5 29 48 PM" src="https://github.com/electron-react-boilerplate/electron-react-boilerplate/assets/13532956/065faeda-3083-4722-840c-2e1ba2d405a7">


Below error resolved by step 2. if we add `sqlite` and `typeorm` in `devDependencies`
<img width="980" alt="Screenshot 2024-04-09 at 5 28 44 PM" src="https://github.com/electron-react-boilerplate/electron-react-boilerplate/assets/13532956/43002e0b-5dca-415a-a5ab-82b5751f9b56">
